patches/server/0032-Optimize-BehaviorController.patch

```diff
@@ -0,0 +1,108 @@
+-                Logger logger = BehaviorController.LOGGER;
++                DataResult<ImmutableList.Builder<BehaviorController.a<?>>> dataresult = mutableobject1.getValue(); // Yatopia - decompile fix
+ 
+-                logger.getClass();
+-                ImmutableList<BehaviorController.a<?>> immutablelist = (ImmutableList) dataresult.resultOrPartial(logger::error).map(Builder::build).orElseGet(ImmutableList::of);
+-                Collection collection2 = collection;
+-                Collection collection3 = collection1;
+-                MutableObject mutableobject2 = mutableobject;
++                ImmutableList<BehaviorController.a<?>> immutablelist = dataresult.resultOrPartial(BehaviorController.LOGGER::error).map(Builder::build).orElseGet(ImmutableList::of); // Yatopia - decompile fix
+ 
+-                mutableobject.getClass();
+-                return DataResult.success(new BehaviorController<>(collection2, collection3, immutablelist, mutableobject2::getValue));
++                return DataResult.success(new BehaviorController<>(collection, collection1, immutablelist, mutableobject::getValue)); // Yatopia - decompile fix
+             }
+ 
+             private <T, U> DataResult<BehaviorController.a<U>> a(MemoryModuleType<U> memorymoduletype, DynamicOps<T> dynamicops, T t0) {
+-                return ((DataResult) memorymoduletype.getSerializer().map(DataResult::success).orElseGet(() -> {
++                return (memorymoduletype.getSerializer().map(DataResult::success).orElseGet(() -> { // Yatopia - decompile fix
+                     return DataResult.error("No codec for memory: " + memorymoduletype);
+                 })).flatMap((codec) -> {
+                     return codec.parse(dynamicops, t0);
+@@ -97,9 +94,9 @@ public class BehaviorController<E extends EntityLiving> {
+             }
+ 
+             public <T> RecordBuilder<T> encode(BehaviorController<E> behaviorcontroller, DynamicOps<T> dynamicops, RecordBuilder<T> recordbuilder) {
+-                behaviorcontroller.j().forEach((behaviorcontroller_a) -> {
++                for (BehaviorController.a<?> behaviorcontroller_a : behaviorcontroller.memoriesList()) { // Yatopia
+                     behaviorcontroller_a.a(dynamicops, recordbuilder);
+-                });
++                } // Yatopia
+                 return recordbuilder;
+             }
+         }).fieldOf("memories").codec());
+@@ -158,6 +155,14 @@ public class BehaviorController<E extends EntityLiving> {
+         return ((Codec) this.b.get()).encodeStart(dynamicops, this);
+     }
+ 
++    // Yatopia start
++    private List<BehaviorController.a<?>> memoriesList() {
++        List<BehaviorController.a<?>> ret = new java.util.ArrayList<>();
++        for (Map.Entry<MemoryModuleType<?>, Optional<? extends ExpirableMemory<?>>> entry : memories.entrySet()) {
++            ret.add(BehaviorController.a.b(entry.getKey(), (Optional) entry.getValue()));
++        }
++        return ret;
++    }
```

Comment:
im tired. please add `// Yatopia end` here when merging or if not merged ill do it tommorrow